prime-router/src/main/kotlin/fhirengine/engine/FHIRTranslator.kt

```diff
@@ -0,0 +1,144 @@
+package gov.cdc.prime.router.fhirengine.engine
+
+import gov.cdc.prime.router.ActionLogger
+import gov.cdc.prime.router.InvalidReportMessage
+import gov.cdc.prime.router.Metadata
+import gov.cdc.prime.router.Options
+import gov.cdc.prime.router.Report
+import gov.cdc.prime.router.SettingsProvider
+import gov.cdc.prime.router.Source
+import gov.cdc.prime.router.azure.ActionHistory
+import gov.cdc.prime.router.azure.BlobAccess
+import gov.cdc.prime.router.azure.DatabaseAccess
+import gov.cdc.prime.router.azure.Event
+import gov.cdc.prime.router.azure.ProcessEvent
+import gov.cdc.prime.router.azure.QueueAccess
+import gov.cdc.prime.router.azure.db.tables.pojos.ItemLineage
+import gov.cdc.prime.router.fhirengine.translation.hl7.FhirToHl7Converter
+import gov.cdc.prime.router.fhirengine.translation.hl7.schema.ConfigSchema
+import gov.cdc.prime.router.fhirengine.utils.FhirTranscoder
+
+/**
+ * Translate a full-ELR FHIR [message] into the formats needed by any receivers from the route step
+ * [metadata] mockable metadata
+ * [settings] mockable settings
+ * [db] mockable database access
+ * [blob] mockable blob storage
+ * [queue] mockable azure queue
+ */
+class FHIRTranslator(
+    metadata: Metadata = Metadata.getInstance(),
+    settings: SettingsProvider = this.settingsProviderSingleton,
+    db: DatabaseAccess = this.databaseAccessSingleton,
+    blob: BlobAccess = BlobAccess(),
+    queue: QueueAccess = QueueAccess,
+) : FHIREngine(metadata, settings, db, blob, queue) {
+
+    /**
+     * Accepts a FHIR [message], parses it, and generates translated output files for each item in the destinations
+     *  element.
+     * [actionHistory] and [actionLogger] ensure all activities are logged.
+     * [metadata] will usually be null; mocked metadata can be passed in for unit tests
+     */
+    override fun doWork(
+        message: RawSubmission,
+        actionLogger: ActionLogger,
+        actionHistory: ActionHistory,
+        metadata: Metadata?
+    ) {
+        logger.trace("Translating FHIR file for receivers.")
+        try {
+            // pull fhir document and parse FHIR document
+            val bundle = FhirTranscoder.decode(message.downloadContent())
+
+            // track input report
+            actionHistory.trackExistingInputReport(message.reportId)
+
+            // todo: iterate over each receiver, translating on a per-receiver basis - for phase 1, hard coded to CO
+            val receivers = listOf<String>("co-pdh")
+
+            receivers.forEach {
+                // todo: get schema for receiver - for Phase 1 this is solely going to convert to HL7 and not do any
+                //  receiver-specific transforms
+
+                // todo: do translation, get hl7 message
+                var schema = ConfigSchema("FHIR_HL7", hl7Type = "ORU_R01", hl7Version = "2.5.1")
```

Comment:
Just FYI: This will change to be loaded from files.

Reply:
Agreed, my understanding is we are not ready to do that yet.